quickwit/quickwit-ingest/src/ingest_v2/fetch.rs

```diff
@@ -0,0 +1,605 @@
+                from_position_inclusive: self.from_position_inclusive,
+            };
+            self.from_position_inclusive += num_docs;
+
+            if self
+                .fetch_response_tx
+                .send(Ok(fetch_response))
+                .await
+                .is_err()
+            {
+                // The ingester was dropped.
+                break;
+            }
+        }
+        debug!(
+            client_id=%self.client_id,
+            index_uid=%self.index_uid,
+            source_id=%self.source_id,
+            shard_id=%self.shard_id,
+            to_position_inclusive=%self.to_position_inclusive,
+            "Fetch task completed."
+        )
+    }
+}
+
+/// Combines multiple fetch streams originating from different ingesters into a single stream. It
+/// tolerates the failure of ingesters and automatically fails over to replica shards.
+pub struct MultiFetchStream {
+    node_id: NodeId,
+    client_id: ClientId,
+    ingester_pool: IngesterPool,
+    fetch_task_handles: HashMap<QueueId, JoinHandle<()>>,
+    fetch_response_rx: mpsc::Receiver<IngestV2Result<FetchResponseV2>>,
+    fetch_response_tx: mpsc::Sender<IngestV2Result<FetchResponseV2>>,
+}
+
+impl MultiFetchStream {
+    pub fn new(node_id: NodeId, client_id: ClientId, ingester_pool: IngesterPool) -> Self {
+        let (fetch_response_tx, fetch_response_rx) = mpsc::channel(3);
+        Self {
+            node_id,
+            client_id,
+            ingester_pool,
+            fetch_task_handles: HashMap::new(),
+            fetch_response_rx,
+            fetch_response_tx,
+        }
+    }
+
+    /// Subscribes to a shard and fails over to the replica if an error occurs.
+    pub async fn subscribe(
+        &mut self,
+        leader_id: NodeId,
+        follower_id: Option<NodeId>,
+        index_uid: IndexUid,
+        source_id: SourceId,
+        shard_id: ShardId,
+        mut from_position_exclusive: Option<u64>,
+        to_position_inclusive: Option<u64>,
+    ) -> IngestV2Result<()> {
+        let queue_id = queue_id(index_uid.as_str(), &source_id, shard_id);
+        let entry = self.fetch_task_handles.entry(queue_id.clone());
+
+        if let Entry::Occupied(_) = entry {
+            return Err(IngestV2Error::Internal(format!(
+                "Stream has already subscribed to shard `{queue_id}`."
+            )));
+        }
+        let (mut preferred_ingester_id, mut failover_ingester_id) =
+            if let Some(follower_id) = follower_id {
+                if leader_id == self.node_id {
+                    (leader_id, Some(follower_id))
+                } else if follower_id == self.node_id {
+                    (follower_id, Some(leader_id))
+                } else if rand::random::<bool>() {
+                    (leader_id, Some(follower_id))
+                } else {
+                    (follower_id, Some(leader_id))
+                }
+            } else {
+                (leader_id, None)
```

Comment:
I don't understand the logic here. What is the intent, espeically of the random stuff.

Reply:
ok i got the logic, maybe we should isolate it in a function though.